File: frontend/src/renderer/lib/block-viewport.test.ts
import { afterEach, describe, expect, it } from "vitest";
import {
	headerSticks,
	isPinned,
	nextBoundary,
	previousBoundary,
	previousTarget,
	topItemFor,
	VIRTUALIZATION_THRESHOLD,
	virtualizationThreshold,
} from "./block-viewport";

const thresholdGlobal = globalThis as typeof globalThis & {
	__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD?: number;
};

const items = [
	{ index: 0, start: 0, size: 100 },
	{ index: 1, start: 100, size: 250 },
	{ index: 2, start: 350, size: 80 },
];

describe("topItemFor", () => {
	it("finds the item spanning the top edge", () => {
		expect(topItemFor(items, 0)?.index).toBe(0);
		expect(topItemFor(items, 99)?.index).toBe(0);
		expect(topItemFor(items, 100)?.index).toBe(1);
		expect(topItemFor(items, 349)?.index).toBe(1);
		expect(topItemFor(items, 350)?.index).toBe(2);
	});

	it("returns nothing past the end of what is rendered", () => {
		expect(topItemFor(items, 430)).toBeUndefined();
	});

	it("returns nothing for an empty window", () => {
		expect(topItemFor([], 0)).toBeUndefined();
	});
});

describe("headerSticks", () => {
	it("sticks for a block shorter than the viewport", () => {
		expect(headerSticks(200, 600)).toBe(true);
	});

	it("sticks for a block exactly as tall as the viewport", () => {
		expect(headerSticks(600, 600)).toBe(true);
	});

	it("does not stick for a block taller than the viewport", () => {
		expect(headerSticks(900, 600)).toBe(false);
	});

	it("does not stick when the viewport has no height yet", () => {
		expect(headerSticks(100, 0)).toBe(false);
	});
});

describe("isPinned", () => {
	it("is pinned at the tail and inside the slack", () => {
		expect(isPinned(400, 1000, 600)).toBe(true);
		expect(isPinned(380, 1000, 600)).toBe(true);
	});

	it("is not pinned once clear of the slack", () => {
		expect(isPinned(300, 1000, 600)).toBe(false);
	});

	it("a list shorter than its viewport is pinned", () => {
		expect(isPinned(0, 200, 600)).toBe(true);
	});
});

describe("boundaries", () => {
	it("steps forward and stops at the last block", () => {
		expect(nextBoundary(0, 3)).toBe(1);
		expect(nextBoundary(2, 3)).toBeUndefined();
	});

	it("steps forward from nothing to the first block", () => {
		expect(nextBoundary(undefined, 3)).toBe(0);
	});

	it("steps back and stops at the first block", () => {
		expect(previousBoundary(2, 3)).toBe(1);
		expect(previousBoundary(0, 3)).toBeUndefined();
	});

	it("has no boundary in an empty list", () => {
		expect(nextBoundary(undefined, 0)).toBeUndefined();
		expect(previousBoundary(0, 0)).toBeUndefined();
	});
});

describe("previousTarget", () => {
	it("returns to the start of a partly scrolled block first", () => {
		expect(previousTarget({ index: 1, start: 100, size: 250 }, 180, 3)).toBe(1);
	});

	it("steps to the block before once already at a boundary", () => {
		expect(previousTarget({ index: 1, start: 100, size: 250 }, 100, 3)).toBe(0);
	});

	it("has nowhere to go from the first block's start", () => {
		expect(previousTarget({ index: 0, start: 0, size: 100 }, 0, 3)).toBeUndefined();
	});

	it("has nowhere to go with no top item", () => {
		expect(previousTarget(undefined, 0, 3)).toBeUndefined();
	});
});

describe("VIRTUALIZATION_THRESHOLD", () => {
	it("is the documented threshold (100)", () => {
		expect(VIRTUALIZATION_THRESHOLD).toBe(100);
	});
});

describe("virtualizationThreshold", () => {
	afterEach(() => {
		delete thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD;
	});

	it("returns the documented constant when the test global is unset", () => {
		expect(virtualizationThreshold()).toBe(VIRTUALIZATION_THRESHOLD);
	});

	it("returns the documented constant when the test global is not a finite number", () => {
		thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD = Number.NaN;
		expect(virtualizationThreshold()).toBe(VIRTUALIZATION_THRESHOLD);

		thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD = Number.POSITIVE_INFINITY;
		expect(virtualizationThreshold()).toBe(VIRTUALIZATION_THRESHOLD);

		thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD = "4" as unknown as number;
		expect(virtualizationThreshold()).toBe(VIRTUALIZATION_THRESHOLD);
	});

	it("returns the override when the test global is a finite number", () => {
		thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD = 4;
		expect(virtualizationThreshold()).toBe(4);

		thresholdGlobal.__OPERATOR_E2E_BLOCK_VIRTUALIZATION_THRESHOLD = 0;
		expect(virtualizationThreshold()).toBe(0);
	});
});

File: packages/mobile/test/feature/blocks/logic/block_viewport_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:operator_mobile/feature/blocks/logic/block_viewport.dart';
import 'package:operator_mobile/feature/blocks/logic/session_block.dart';

SessionBlock _block(int seq) => SessionBlock(
  id: 'seq-$seq',
  firstSeq: seq,
  lastSeq: seq,
  kind: BlockKind.tool,
  status: BlockStatus.ok,
  title: 'Bash',
  body: 'ok',
);

void main() {
  group('pivotIndex', () {
    test('an unset pivot puts every block after the centre', () {
      expect(BlockViewport.pivotIndex([_block(4), _block(5)], null), 0);
    });

    test('splits older blocks out of the centre sliver', () {
      final blocks = [_block(1), _block(2), _block(3), _block(4)];
      expect(BlockViewport.pivotIndex(blocks, 3), 2);
    });

    test(
      'a pivot older than everything held leaves the leading sliver empty',
      () {
        expect(BlockViewport.pivotIndex([_block(7), _block(8)], 2), 0);
      },
    );

    test(
      'a pivot newer than everything held puts every block before the centre',
      () {
        expect(BlockViewport.pivotIndex([_block(1), _block(2)], 9), 2);
      },
    );

    test('an evicted pivot still splits at the same seq boundary', () {
      final afterEviction = [_block(3), _block(4), _block(5)];
      expect(BlockViewport.pivotIndex(afterEviction, 4), 1);
    });

    test('an empty window has nothing before the centre', () {
      expect(BlockViewport.pivotIndex(const [], 4), 0);
    });
  });

  group('isPinned', () {
    test('is pinned exactly at the tail', () {
      expect(BlockViewport.isPinned(1000, 1000), isTrue);
    });

    test('is pinned inside the slack', () {
      expect(BlockViewport.isPinned(980, 1000), isTrue);
    });

    test('is not pinned once scrolled clear of the slack', () {
      expect(BlockViewport.isPinned(900, 1000), isFalse);
    });

    test('an unscrollable list is pinned', () {
      expect(BlockViewport.isPinned(0, 0), isTrue);
    });

    test('a negative offset in the leading sliver is not pinned', () {
      expect(BlockViewport.isPinned(-400, 1000), isFalse);
    });
  });

  group('headerSticks', () {
    test('a block shorter than the viewport keeps its header pinned', () {
      expect(BlockViewport.headerSticks(200, 600), isTrue);
    });

    test('a block exactly as tall as the viewport still sticks', () {
      expect(BlockViewport.headerSticks(600, 600), isTrue);
    });

    test('a block taller than the viewport does not trap its own header', () {
      expect(BlockViewport.headerSticks(900, 600), isFalse);
    });
  });

  group('boundaries', () {
    test('next steps forward and stops at the last block', () {
      expect(BlockViewport.nextBoundary(0, 3), 1);
      expect(BlockViewport.nextBoundary(2, 3), isNull);
    });

    test('next from nothing selects the first block', () {
      expect(BlockViewport.nextBoundary(null, 3), 0);
    });

    test('previous steps back and stops at the first block', () {
      expect(BlockViewport.previousBoundary(2, 3), 1);
      expect(BlockViewport.previousBoundary(0, 3), isNull);
    });

    test('an empty list has no boundaries in either direction', () {
      expect(BlockViewport.nextBoundary(null, 0), isNull);
      expect(BlockViewport.previousBoundary(0, 0), isNull);
    });
  });
}
